### PR TITLE
[+] Core : Implement Module Manager

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
_Description coming soon_

### CODE REVIEW

File: src/Core/Addon/Module/ModuleManager.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    /**
+     * Add new module from zipball. This will unzip the file and move the content
+     * to the right locations.
+     * A theme can bundle modules, resources, docuementation, email templates and so on.
+     *
+     * @param string $source The source can be a module name (installed from either local disk or addons.prestashop.com).
+     * or a location (url or path to the zip file)
+     * @return bool true for success
+     */
+    public function install($name)
```
Comment (tchauviere) [inline src/Core/Addon/Module/ModuleManager.php]: Maybe name $name ---> $source to follow PHPDoc ?

File: src/Core/Addon/Module/ModuleManager.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    {
+        if (!$this->employee->can('add', 'AdminModules')) {
+            throw new Exception('You are not allowed to install a module');
+        }
+
+        if ($this->moduleProvider->isInstalled($name)) {
+            throw new Exception('This module is already installed');
```
Comment (maximebiloe) [inline src/Core/Addon/Module/ModuleManager.php]: Maybe it would be usefull to add he module name into the message ?

Comment (tchauviere) [inline src/Core/Addon/Module/ModuleManager.php]: I'm @tchauviere, and I agree with this message !

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 3]
```diff
@@ -209,33 +186,37 @@ public function getCatalogCategories()
         return $this->catalog_categories;
     }
 
-    public function getManageModules(array $filter = [])
+    public function getCategoriesFromModules(&$modules)
```
Comment (Quetzacoalt91) [inline src/Adapter/Module/AdminModuleDataProvider.php]: Does `&` can be removed ?

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 4]
```diff
@@ -209,33 +186,37 @@ public function getCatalogCategories()
...
+                        $name
+                    );
+                }
 
-    public function getModule($name)
-    {
-        if ($this->isModuleOnDisk($name)) {
-            return \PrestaShop\PrestaShop\Adapter\ServiceLocator::get($name);
+                $categories['categories']->subMenu[$ref]->modulesRef[] = $module->attributes->get('name');
+                $module->refs[$key] = $ref;
```
Comment (Quetzacoalt91) [inline src/Adapter/Module/AdminModuleDataProvider.php]: `$module->attributes->replace('refs', ($module->attributes->get('refs'))[$key] = $ref;` ?

File: src/Adapter/Module/Module.php [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            return false;
+        }
+        return $this->instance->reset();
+    }
+
+    protected function instanciateLegacyModule()
+    {
+        require_once _PS_MODULE_DIR_ . DIRECTORY_SEPARATOR . $this->attributes->get('name') . DIRECTORY_SEPARATOR . $this->attributes->get('name') . '.php';
+        $this->instance = \Module::getInstanceByName($this->attributes->get('name'));
+        $this->reflector = new ReflectionObject($this->instance);
```
Comment (Quetzacoalt91) [inline src/Adapter/Module/Module.php]: To remove, unused

File: src/Adapter/Module/ModuleDataUpdater.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+class ModuleDataUpdater
+{
+
+    public function setModuleOnDiskFromAddons($name)
+    {
+        // Note : Data caching should be handled by the addons data provider
+
+        $addons_provider = new AddonsDataProvider();
+        // Check if the module can be downloaded from addons
+        foreach ((new AdminModuleDataProvider())->getCatalogModules() as $catalog_module) {
```
Comment (Quetzacoalt91) [inline src/Adapter/Module/ModuleDataUpdater.php]: `new AdminModuleDataProvider()` should be stored in a variable

File: src/Core/Addon/AddonListFilterType.php [hunk 7]
```diff
@@ -28,7 +28,10 @@
 
 class AddonListFilterType
 {
+    /* Bitwise operator */
     const THEME = 1;
     const MODULE = 2;
-    const ALL = 3;
+    const SERVICE = 4;
+
+    const ALL = 7; /* = 1 & 2 & 4 */
```
Comment (Quetzacoalt91) [inline src/Core/Addon/AddonListFilterType.php]:
```php
const ALL = 7; /* = 1 | 2 | 4 */
```

File: src/Core/Addon/Module/ModuleRepository.php [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+     *
+     * @param  string $name The technical module name to find
+     * @return array         Module data stored in file
+     */
+    private function generateCacheFile($data)
+    {
+        $yml_file_path = self::CACHE_FILE; //_PS_MODULE_DIR_.$name.'/'.$name.'.yml';
+        /* $dumper = new Dumper();
+
+          $yaml = $dumper->dump($data, 3); */
```
Comment (Quetzacoalt91) [inline src/Core/Addon/Module/ModuleRepository.php]: Clean up to do

File: src/PrestaShopBundle/Controller/Admin/ModuleController.php [hunk 9]
```diff
@@ -264,8 +290,19 @@ public function importModuleAction(Request $request)
...
-            $installation_response['module_name'] = $module_name;
+            $installation_response = [
+                'status' => $moduleManager->install($module_name),
+                'msg' => '',
+                'module_name' => $module_name,
+            ];
+
+            if ($installation_response['status'] === null) {
+                $installation_response['status'] = false;
+                $installation_response['msg'] = $module_name .' did not returned a valid response on install action';
```
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Controller/Admin/ModuleController.php]: **did not return**

File: src/PrestaShopBundle/Controller/Admin/ModuleController.php [hunk 10]
```diff
@@ -348,18 +364,8 @@ protected function getToolbarButtons()
         return $toolbarButtons;
     }
 
-    final private function getTopMenuData($source = 'catalog', $activeMenu = null)
+    final private function getTopMenuData(array $topMenuData, $activeMenu = null)
```
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Controller/Admin/ModuleController.php]: `final` to remove

File: src/PrestaShopBundle/Controller/Admin/ModuleController.php [hunk 11]
```diff
@@ -430,96 +388,6 @@ public function configureModuleAction($module_name)
...
-                $msg = sprintf('Could not update module %s, failed while applying upgrade file %s. The module has been disabled.', $module_name, $details['version_fail']);
-            } else {
-                $msg = sprintf('Could not update module %s.', $module_name);
-            }
-        }
-
-        return array('status' => $status, 'msg' => $msg);
-    }
-
     final private function getAddonsConnectToolbar()
```
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Controller/Admin/ModuleController.php]: `final` to remove

File: admin-dev/themes/default/js/bundle/module/module.js [hunk 12]
```diff
@@ -288,6 +295,7 @@ var AdminModule = function () {
                // Added timeout for a better render of animation and avoid to have displayed at the same time
                setTimeout(function() {
                    $(_these.moduleImportStartSelector).fadeIn(function(event){
+                       $(_these.moduleImportFailureMsgDetailsSelector).css('display', 'none');
```
Comment (mickaelandrieu) [inline admin-dev/themes/default/js/bundle/module/module.js]: use `hide()` instead.

File: admin-dev/themes/default/js/bundle/module/module.js [hunk 13]
```diff
@@ -964,12 +981,12 @@ var AdminModule = function () {
...
+                $(this).addClass('module-item-list col-lg-12');
                 _this.setNewDisplay($(this), '-grid', '-list');
             });
             // Change module addons item
             addonItem.removeClass();
-            addonItem.addClass('module-addons-item-list col-md-12');
+            addonItem.addClass('module-addons-item-list col-lg-12');
             this.setNewDisplay(addonItem, '-grid', '-list');
         } else {
             console.error('Can\'t switch to undefined display property "' + switchTo + '"');
```
Comment (mickaelandrieu) [inline admin-dev/themes/default/js/bundle/module/module.js]: What do you think about rename `switchTo` to `displayProperty` ?

File: install-dev/models/install.php [hunk 14]
```diff
@@ -814,18 +815,18 @@ public function installModules($module = null)
...
                 continue;
             }
 
-            $module = Module::getInstanceByName($module_name);
-            if (!$module->install()) {
-                $module_errors = $module->getErrors();
-                if (empty($module_errors)) {
+            if (!$moduleManager->install($module_name)) {
+                /*$module_errors = $module->getErrors();
+                if (empty($module_errors)) {*/
```
Comment (mickaelandrieu) [inline install-dev/models/install.php]: is that intended ?

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 15]
```diff
@@ -48,9 +48,6 @@ class AdminModuleDataProvider extends AbstractAdminQueryBuilder implements Modul
     const _CACHEFILE_CATEGORIES_ = 'catalog_categories.json';
     const _CACHEFILE_MODULES_ = 'catalog_modules.json';
 
-    const _CACHEFILE_INSTALLED_CATEGORIES_ = 'catalog_installed_categories.json';
-    const _CACHEFILE_INSTALLED_MODULES_ = 'catalog_installed_modules.json';
-
     /* Cache for One Day */
     const _WATCH_DOG_ = 86400;
```
Comment (mickaelandrieu) [inline src/Adapter/Module/AdminModuleDataProvider.php]: what do you think about rename it `_DAY_IN_SECONDS_` ?

File: src/Adapter/Module/AdminModuleDataProvider.php [hunk 16]
```diff
@@ -611,15 +453,6 @@ protected function fallbackOnCatalogCache()
...
-    {
-        // Fallback on data from cache if exists
-        $this->manage_categories = $this->getModuleCache(self::_CACHEFILE_INSTALLED_CATEGORIES_, false);
-        $this->manage_modules    = $this->getModuleCache(self::_CACHEFILE_INSTALLED_MODULES_, false);
-
-        return ($this->manage_categories && $this->manage_modules);
-    }
-
     private function getModuleCache($file, $check_freshness = true)
     {
```
Comment (mickaelandrieu) [inline src/Adapter/Module/AdminModuleDataProvider.php]: I need more time to review this file ... I think too much responsabilities for only one class, did you take a look at wha

Comment (mickaelandrieu) [inline src/Adapter/Module/AdminModuleDataProvider.php]: I need more time to review this file ... I think too much responsabilities for only one class, did you take a look at the Theme implementation with Manager and Repository ?

File: src/Adapter/Module/ModuleDataProvider.php [hunk 17]
```diff
@@ -25,20 +25,108 @@
...
 namespace PrestaShop\PrestaShop\Adapter\Module;
 
-/**
- * This class will provide data from DB / ORM about Module
- */
+use PrestaShop\PrestaShop\Adapter\Shop\Context;
+
 class ModuleDataProvider
 {
+    public function findByName($name)
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleDataProvider.php]: this class should be called ModuleRepository regarding the consistency with Theme implementation.

File: src/Adapter/Module/ModuleDataUpdater.php (hunk 6 above)
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleDataUpdater.php]: a lot of code is already incompatible with PHP 5.4, I suggest to decide first what are the minimal requirements and then do the work to make the code compliant with, what do you think ?

File: src/Core/Addon/Module/ModuleManager.php [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+        if (!$this->employee->can('edit', 'AdminModules')
+            || !$this->moduleProvider->can('configure', $name)) {
+            throw new Exception('You are not allowed to upgrade this module');
+        }
+
+        if (! $this->moduleProvider->isInstalled($name)) {
+            throw new Exception('This module must be installed');
+        }
+
+        // Get new module
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleManager.php]: I'd like to see this implementation complete before merge. We have a lot (too much ?) code merged and still unfinished, this lead to a lot of debugging.

If you can't finish it, throw an `NotImplementedException` here please.

File: src/Core/Addon/Module/ModuleRepository.php [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+        $yaml          = json_encode($data);
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleRepository.php]: I'm pretty sure you won't receive any YAML atm :dancers:

File: src/Core/Addon/Module/ModuleRepository.php [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+    /**
+     * We load the file which contains cached data
+     *
+     * @return array         Module data loaded in file
+     */
+    private function readCacheFile()
+    {
+        $file_path = self::CACHE_FILE;
+
+        // YML file not found ? Generate it
```
Comment (mickaelandrieu) [inline src/Core/Addon/Module/ModuleRepository.php]: Json or Yaml ?

File: src/PrestaShopBundle/Controller/Admin/ModuleController.php [hunk 21]
```diff
@@ -18,12 +21,21 @@ class ModuleController extends FrameworkBundleAdminController
      */
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ModuleController.php]: ModuleRepository could be (should be ?) declared as a service.

File: src/PrestaShopBundle/Controller/Admin/ModuleController.php [hunk 22]
```diff
@@ -136,14 +152,24 @@ public function manageAction(Request $request, $category = null, $keyword = null
 
     public function moduleAction(Request $request)
     {
-        $action = $request->attributes->get('action'). 'Module';
+        $action = $request->attributes->get('action');
         $module = $request->attributes->get('module_name');
+        $moduleManager = (new ModuleManagerBuilder())->build();
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ModuleController.php]: ModuleManager should be declared as a service too

File: src/PrestaShopBundle/Resources/views/Admin/Module/_partials/_modules_modal_read_more.html.twig [hunk 23]
```diff
@@ -1,15 +1,15 @@
-<div id="module-modal-read-more-{{module.name}}{{ additionalModalSuffix|default('') }}" class="modal modal-vcenter fade" role="dialog">
+<div id="module-modal-read-more-{{module.attributes.get('name')}}{{ additionalModalSuffix|default('') }}" class="modal modal-vcenter fade" role="dialog">
     <div class="modal-dialog module-modal-dialog">
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/Module/_partials/_modules_modal_read_more.html.twig]: How about a `ModulePresenter` able to return exploitable representation in front ?

**Comment by mickaelandrieu** [comment]
Tests need to be done, by the way can't way to be able to use this implementation :)

But... how about the "sandbox" ?I don't retrieve a `safety` mechanism to disable module in case of Exception.

File: src/Adapter/Module/ModuleDataUpdater.php (hunk 6 above)
Comment (Quetzacoalt91) [inline src/Adapter/Module/ModuleDataUpdater.php]: See last mail from @Shudrum. The code must be compliant with PHP 5.4 :wink:

Comment (Shudrum) [inline src/Adapter/Module/ModuleDataUpdater.php]: @mickaelandrieu, we have made a CR yesterday, don't worry :)

Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleDataUpdater.php]: Ok :)

File: src/PrestaShopBundle/Resources/views/Admin/Module/_partials/_modules_modal_read_more.html.twig (hunk 23 above)
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Resources/views/Admin/Module/_partials/_modules_modal_read_more.html.twig]: Interesting idea, we will introduce this new class later as an improvement !

**Comment by maximebiloe** [comment]
giphy

**Pull request merged**